pydicom/pixel_data_handlers/pillow_handler.py

```diff
@@ -159,70 +160,48 @@ def get_pixeldata(ds):
             im.draft('YCbCr', (ds.Rows, ds.Columns))
         pixel_bytes.extend(im.tobytes())
 
-        j2k_precision = _get_j2k_precision(pixel_data)
+        params = get_j2k_parameters(pixel_data)
+        j2k_precision = params.setdefault("precision", ds.BitsStored)
+        j2k_sign = params.setdefault("is_signed", None)
 
-    logger.debug("Successfully read %s pixel bytes", len(pixel_bytes))
+    logger.debug(f"Successfully read {len(pixel_bytes)} pixel bytes")
 
     arr = numpy.frombuffer(pixel_bytes, pixel_dtype(ds))
 
     if transfer_syntax in PillowJPEG2000TransferSyntaxes:
-        # Pillow converts N-bit data to 8- or 16-bit unsigned data
+        # Pillow converts N-bit data to 8- or 16-bit unsigned data,
         # See Pillow src/libImaging/Jpeg2KDecode.c::j2ku_gray_i
-        if ds.PixelRepresentation == 1:
-            # Pillow converts signed data to unsigned
-            #   so we need to undo this conversion
-            arr -= 2**(ds.BitsAllocated - 1)
-
+        shift = ds.BitsAllocated - ds.BitsStored
         if j2k_precision and j2k_precision != ds.BitsStored:
             warnings.warn(
-                "The (0028,0101) 'Bits Stored' value doesn't match the "
-                "sample bit depth of the JPEG2000 pixel data ({} vs {} bit). "
-                "It's recommended that you first change the 'Bits Stored' "
-                "value to match the JPEG2000 bit depth in order to get the "
-                "correct pixel data".format(ds.BitsStored, j2k_precision)
+                f"The (0028,0101) 'Bits Stored' value ({ds.BitsStored}-bit) "
+                f"doesn't match the JPEG 2000 data ({j2k_precision}-bit). "
+                f"It's recommended that you change the 'Bits Stored' value"
             )
 
-        shift = ds.BitsAllocated - ds.BitsStored
-        if shift:
-            logger.debug("Shifting right by {} bits".format(shift))
-            numpy.right_shift(arr, shift, out=arr)
+        if config.APPLY_J2K_CORRECTIONS and j2k_precision:
+            # Corrections based on J2K data
+            shift = ds.BitsAllocated - j2k_precision
+            if not j2k_sign and j2k_sign != ds.PixelRepresentation:
+                # Convert unsigned J2K data to 2's complement
+                numpy.right_shift(arr, shift, out=arr)
+            else:
+                if ds.PixelRepresentation == 1:
+                    # Pillow converts signed data to unsigned
+                    #   so we need to undo this conversion
+                    arr -= 2**(ds.BitsAllocated - 1)
+
+                if shift:
+                    numpy.right_shift(arr, shift, out=arr)
+        else:
+            # Corrections based on dataset elements
+            if ds.PixelRepresentation == 1:
+                arr -= 2**(ds.BitsAllocated - 1)
+
+            if shift:
+                numpy.right_shift(arr, shift, out=arr)
 
```

Comment:
Is it guaranteed safe to use the same array as `out`?  Seems like it might be modifying it in place while the operation is underway.

Reply:
I've never seen anything in the numpy documentation that indicates its not safe. It's a pretty standard practice when you can reuse the memory allocated to an array.

Reply:
A bit more reading suggests that doing this is probably a bit slower for this use case